# Text Analytics - Document Retrieval

In [1]:
import graphlab

In [4]:
people = graphlab.SFrame('people_wiki.gl/')
people.head()

This non-commercial license of GraphLab Create for academic use is assigned to snehchem@iu.edu and will expire on December 16, 2017.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: C:\Users\sneha\AppData\Local\Temp\graphlab_server_1482248842.log.0


Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

In [5]:
people['word_count'] = graphlab.text_analytics.count_words(people['text'])

In [6]:
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'since': 1L, 'carltons': ... |
| alfred j lewy aka sandy le... | {'precise': 1L, 'thomas': ... |
| harpdog brown is a singer ... | {'just': 1L, 'issued': 1L,... |
| franz rottensteiner born i... | {'all': 1L, 'bauforschung'... |
| henry krvits born 30 decem... | {'legendary': 1L, 'gangste... |
| sam henderson born october... | {'now': 1L, 'currently': 1... |
| aaron lacrate is an americ... | {'exclusive': 2L, 'produce... |
| trevor ferguson aka john f... | {'taxi': 1L, 'salon': 1L, ... |
| grant nelson born 27 april... | {'houston': 1L, 'frankie':... |
| cathy caruth born 1955 is ... | {'phenomenon': 1L, 'debora... |
+-------------------------------+-------------------------------+
[10 rows x 4 columns]

# 1. Compare top words according to word counts to TF-IDF: 

## Now, take a particular famous person, 'Elton John'. What are the 3 words in his articles with highest word counts? What are the 3 words in his articles with highest TF-IDF? 

## These results illustrate why TF-IDF is useful for finding important words.

In [7]:
tfidf = graphlab.text_analytics.tf_idf(people['word_count'])
people['tfidf'] = tfidf

In [8]:
elton = people[people['name'] == 'Elton John']
elton

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'all': 1L, 'six': 1L, 'pr... | {'all': 1.6431112434912472... |
+-------------------------------+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [9]:
elton[['tfidf']].stack('tfidf',new_column_name=['word','tfidf']).sort('tfidf',ascending=False)

Columns:
	word	str
	tfidf	float

Rows: 255

Data:
+---------------+---------------+
|      word     |     tfidf     |
+---------------+---------------+
|    furnish    |  18.38947184  |
|     elton     |  17.48232027  |
|   billboard   | 17.3036809575 |
|      john     | 13.9393127924 |
|  songwriters  |  11.250406447 |
|  overallelton | 10.9864953892 |
| tonightcandle | 10.9864953892 |
|    19702000   | 10.2933482087 |
|   fivedecade  | 10.2933482087 |
|      aids     |  10.262846934 |
+---------------+---------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [13]:
elton[['word_count']].stack('word_count',new_column_name=['word','count']).sort('count',ascending=False)

Columns:
	word	str
	count	int

Rows: 255

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   27  |
|   in  |   18  |
|  and  |   15  |
|   of  |   13  |
|   a   |   10  |
|  has  |   9   |
|   he  |   7   |
|  john |   7   |
|   on  |   6   |
| since |   5   |
+-------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## the 3 words in his articles with highest word counts are The,In,And
## the 3 words in his articles with highest TF-IDF are furnish, elton, billboard

## 2. Measuring distance: Elton John is a famous singer; let’s compute the distance between his article and those of two other famous singers.
 
## Calculate the cosine distance between the articles on ‘Elton John’ and ‘Victoria Beckham’
## Calculate the cosine distance between the articles on ‘Elton John’ and Paul McCartney’? 
## Which one of the two is closest to Elton John? Does this result make sense?

In [10]:
victoria = people[people['name'] == 'Victoria Beckham']
paul = people[people['name'] == 'Paul McCartney']

In [11]:
graphlab.distances.cosine(elton['tfidf'][0],victoria['tfidf'][0])

0.9567006376655429

In [12]:
graphlab.distances.cosine(elton['tfidf'][0],paul['tfidf'][0])

0.8250310029221779

# Elton John is closer to Paul McCartney -- 0.8250310029221779

## 3. Building nearest neighbors models with different input features and setting the distance metric: 
## In the sample notebook, we built a nearest neighbors model for retrieving articles using TF-IDF as features and using the default setting in the construction of the nearest neighbors model. 
## Here we built two nearest neighbors models:
### (a) Using word counts as features
### (b) Using TF-IDF as features
## In both of these models, we are going to set the distance function to cosine similarity. 

In [16]:
knn_model_tfidf = graphlab.nearest_neighbors.create(people,features=['tfidf'],label='name',distance ='cosine')

Starting brute force nearest neighbors model training.

In [17]:
knn_model_word_count = graphlab.nearest_neighbors.create(people,features=['word_count'],label='name',
                                                    distance ='cosine')

Starting brute force nearest neighbors model training.

In [18]:
# What’s the most similar article, other than itself, to the one on ‘Elton John’ using word count features?
knn_model_word_count.query(elton)  # Ans- Cliff Richard 0.16142415259

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 15.628ms     |

| Done         |         | 100         | 250.029ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------+-------------------+------+
| query_label |  reference_label  |      distance     | rank |
+-------------+-------------------+-------------------+------+
|      0      |     Elton John    | 2.22044604925e-16 |  1   |
|      0      |   Cliff Richard   |   0.16142415259   |  2   |
|      0      |   Sandro Petrone  |   0.16822542751   |  3   |
|      0      |    Rod Stewart    |   0.168327165587  |  4   |
|      0      | Malachi O'Doherty |   0.177315545979  |  5   |
+-------------+-------------------+-------------------+------+
[5 rows x 4 columns]

In [19]:
# What’s the most similar article, other than itself, to the one on ‘Elton John’ using TF-IDF features?
knn_model_tfidf.query(elton)  # Ans- Rod Stewart 0.717219667893

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 15.627ms     |

| Done         |         | 100         | 324.144ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+--------------------+------+
| query_label | reference_label  |      distance      | rank |
+-------------+------------------+--------------------+------+
|      0      |    Elton John    | -2.22044604925e-16 |  1   |
|      0      |   Rod Stewart    |   0.717219667893   |  2   |
|      0      |  George Michael  |   0.747600998969   |  3   |
|      0      | Sting (musician) |   0.747671954431   |  4   |
|      0      |   Phil Collins   |   0.75119324879    |  5   |
+-------------+------------------+--------------------+------+
[5 rows x 4 columns]

In [20]:
# What’s the most similar article, other than itself, to the one on ‘Victoria Beckham’ using word count features?

In [21]:
knn_model_word_count.query(victoria) # Ans - Mary Fitzgerald (artist) 0.207307036115

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 0us          |

| Done         |         | 100         | 253.032ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------------+--------------------+------+
| query_label |     reference_label      |      distance      | rank |
+-------------+--------------------------+--------------------+------+
|      0      |     Victoria Beckham     | -2.22044604925e-16 |  1   |
|      0      | Mary Fitzgerald (artist) |   0.207307036115   |  2   |
|      0      |      Adrienne Corri      |   0.214509782788   |  3   |
|      0      |     Beverly Jane Fry     |   0.217466468741   |  4   |
|      0      |      Raman Mundair       |   0.217695474992   |  5   |
+-------------+--------------------------+--------------------+------+
[5 rows x 4 columns]

In [22]:
# What’s the most similar article, other than itself, to the one on ‘Victoria Beckham’ using TF-IDF features?
knn_model_tfidf.query(victoria)  # Ans - David Beckham 0.548169610263

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 15.626ms     |

| Done         |         | 100         | 359.413ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+---------------------+-------------------+------+
| query_label |   reference_label   |      distance     | rank |
+-------------+---------------------+-------------------+------+
|      0      |   Victoria Beckham  | 1.11022302463e-16 |  1   |
|      0      |    David Beckham    |   0.548169610263  |  2   |
|      0      | Stephen Dow Beckham |   0.784986706828  |  3   |
|      0      |        Mel B        |   0.809585523409  |  4   |
|      0      |    Caroline Rush    |   0.819826422919  |  5   |
+-------------+---------------------+-------------------+------+
[5 rows x 4 columns]